### PR TITLE
Include overwrite data option in SimulationDataSource 

### PR DESCRIPTION
**Pull request by flongford**
This PR introduces an option on the `SimulationDataSourceModel` class to overwrite any simulation data if it already exists on file. Consequently, a new check is performed before running the simulation to see whether 

- `A`: an output data file already exists
- `B`: the `SimulationDataSourceModel` requires this data to be overwritten

If `A` returns `False`, then a simulation is always performed. 

If `A` returns `True`, then a simulation will only be performed when `B` returns `True` also.

Whilst implementing this check, we have refactored some of the required functionalities of `GromacsSimulationBuilder` into the new interface class `ISimualtionBuilder`. This can be considered a step towards obataining generic simulation classes for the BDSS, similar to #14.

### Change log:

- `SimulationDataSourceModel` class includes `ow_data` attribute
- New `ISimulationBuilder` interface class requires subclasses to implement a `get_results_path` method, which will define output file location for simulation data.
- `GromacsSimulationBuilder` class now provides `ISimulationBuilder` class, and so can be queried to provide an output data file path before a simulation is performed.
- `SimulationDataSource` class includes context manager block using `_check_perform_simulation` method to return outcome of `A` and `B` checks for an `GromacsSimulationBuilder` instance
- Plugin no longer contributes `SimulationFactory` to be usable from the BDSS, since the `SimulationDataSource` requires the `create_simulation_builder` method to be implemented

### CODE REVIEW

File: force_gromacs/data_sources/simulation/simulation_data_source.py [hunk 1]
```diff
@@ -18,6 +20,15 @@ class SimulationDataSource(BaseDataSource):
...
     simulation_builder = Instance(GromacsSimulationBuilder)
 
+    def _check_perform_simulation(self, model, results_path):
+        """Check to see whether a simulation should be performed.
+        If a results file already exists, then only perform a new
+        simulation if required by the model"""
+
+        if os.path.exists(results_path):
+            return model.ow_data
+        return True
```
Comment (Corwinpro) [inline force_gromacs/data_sources/simulation/simulation_data_source.py:28]: What if the result file doesn't exist, but the model says "no simulation"?

File: force_gromacs/data_sources/simulation/simulation_model.py [hunk 2]
```diff
@@ -25,6 +25,9 @@ class SimulationDataSourceModel(BaseDataSourceModel):
     #: Total number of fragments in simulation
     size = Int(1000)
 
+    #: Whether or not to overwrite existing simulation data
+    ow_data = Bool(False)
```
Comment (Corwinpro) [inline force_gromacs/data_sources/simulation/simulation_model.py:29]: Does the user see "Ow data" as the label in the UI? Shall we add a more informative label then, like "Overwrite simulation data"?

**Review by Corwinpro — CHANGES_REQUESTED** [review]
Looks great, I would only double check the overwrite things I mentioned in the comments (when we want to overwrite, and how does the user see the UI label). Happy to merge then, if you think we want to keep some methods as  NotImplemented for now.

File: force_gromacs/data_sources/simulation/simulation_model.py (hunk 2 above)
Comment (flongford) [reply force_gromacs/data_sources/simulation/simulation_model.py:29]: That's a good idea, the `ow_data` attribute should already be included in the traits view, but I haven't added a custom label for it yet.

File: force_gromacs/data_sources/simulation/simulation_data_source.py (hunk 1 above)
Comment (flongford) [reply force_gromacs/data_sources/simulation/simulation_data_source.py:28]: Good question, if `model.dry_run==True`, then this is currently dealt with later on by the `GromacsPipeline` class. In which case the directory tree for the simulation data is also created and a bash script produced, but the `GromacsPipeline` does not call any executable commands.

This also means that a result file will not be produced, and so may cause a problem for any data source attempting to performing post-processing operations if the file does not exist.

**Comment by flongford** [comment]
>  Happy to merge then, if you think we want to keep some methods as NotImplemented for now.

Sure - similar to some objects in the `force-bdss`, the intention is for these objects to act as base classes for developers to implement. Do you think therefore it would be clearer to rename the class containing these methods to `BaseGromacsSimulationBuilder`?

**Comment by Corwinpro** [comment]
> > Happy to merge then, if you think we want to keep some methods as NotImplemented for now.
> 
> Sure - similar to some objects in the `force-bdss`, the intention is for these objects to act as base classes for developers to implement. Do you think therefore it would be clearer to rename the class containing these methods to `BaseGromacsSimulationBuilder`?

Good question, I didn't think about it - I thought that the `GromacsSimulationBuilder` is a ready to use class. How different can the NotImplemented methods be?

File: force_gromacs/gromacs_plugin.py [hunk 3]
```diff
@@ -39,6 +39,5 @@ def get_factory_classes(self):
         return [
             FragmentFactory,
             MoleculeFactory,
-            SimulationFactory,
```
Comment (flongford) [inline force_gromacs/gromacs_plugin.py:42]: I've removed the `SimulationFactory` from the plugin, as the `SimulationDataSource` contains methods that need to be implemented (and so cannot be run 'out-of-the-box')

**Comment by flongford** [comment]
> > > Happy to merge then, if you think we want to keep some methods as NotImplemented for now.
> > 
> > 
> > Sure - similar to some objects in the `force-bdss`, the intention is for these objects to act as base classes for developers to implement. Do you think therefore it would be clearer to rename the class containing these methods to `BaseGromacsSimulationBuilder`?
> 
> Good question, I didn't think about it - I thought that the `GromacsSimulationBuilder` is a ready to use class. How different can the NotImplemented methods be?

They have a fixed input / output signature, but otherwise are pretty generic - we could probably think of a better way to handle this using defaults but that might be out of the scope of the current PR

**Review by Corwinpro — APPROVED** [review]
Great, I think it is good to go. LGTM.

**Pull request merged**